pkg/logql/probabilistic_evaluator.go

```diff
@@ -0,0 +1,459 @@
+package logql
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"sort"
+
+	"strconv"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/opentracing/opentracing-go"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql"
+	promql_parser "github.com/prometheus/prometheus/promql/parser"
+
+	"github.com/grafana/dskit/tenant"
+
+	"github.com/grafana/loki/pkg/logql/sketch"
+	"github.com/grafana/loki/pkg/logql/syntax"
+	"github.com/grafana/loki/pkg/logqlmodel"
+	"github.com/grafana/loki/pkg/util"
+	"github.com/grafana/loki/pkg/util/validation"
+)
+
+type T int64
+
+const (
+	VecType T = iota
+	TopKVecType
+)
+
+type StepResult interface {
+	Type() T
+
+	SampleVector() promql.Vector
+	TopkVector() sketch.TopKVector
+}
+
+type SampleVector promql.Vector
+
+func (SampleVector) Type() T {
+	return VecType
+}
+
+func (s SampleVector) SampleVector() promql.Vector {
+	return promql.Vector(s)
+}
+
+func (s SampleVector) TopkVector() sketch.TopKVector {
+	return sketch.TopKVector{}
+}
+
+type TopKVector sketch.TopKVector
+
+func (TopKVector) Type() T {
+	return TopKVecType
+}
+
+func (v TopKVector) SampleVector() promql.Vector {
+	return nil
+}
+
+func (v TopKVector) TopkVector() sketch.TopKVector {
+	return sketch.TopKVector(v)
+}
+
+// ProbabilisticStepEvaluator evaluate a single step of a probabilistic query type.
+type ProbabilisticStepEvaluator interface {
+	// while Next returns a StepResult.
+	Next() (ok bool, ts int64, r StepResult)
+	// Close all resources used.
+	Close() error
+	// Reports any error
+	Error() error
+	// the type of a step result
+	Type() T
+}
+
+type ProbabilisticEvaluator struct {
+	Evaluator // this evaluator should be a DefaultEvaluator
+	logger    log.Logger
+}
+
+type pTopkStepEvaluator struct {
+	expr   *syntax.VectorAggregationExpr
+	next   StepEvaluator
+	logger log.Logger
+}
+
+func (e *pTopkStepEvaluator) Type() T {
+	return TopKVecType
+}
+
+func (e *pTopkStepEvaluator) Next() (bool, int64, StepResult) {
+	next, ts, vec := e.next.Next()
+
+	if !next {
+		return false, 0, nil
+	}
+	if e.expr.Params < 1 {
+		return next, ts, nil
+	}
+
+	// We only use one aggregation. The topk sketch compresses all
+	// information and thus we don't need to take care of grouping
+	// here.
+	topkAggregation, err := sketch.NewCMSTopkForCardinality(e.logger, 10, 100000)
+	if err != nil {
+		// TODO(karsten): capture error
+		return false, ts, nil
+	}
+
+	buf := make([]byte, 0, 1024)
+	sort.Strings(e.expr.Grouping.Groups)
+
+	for _, s := range vec {
+		metric := s.Metric
+		
+		var groupingKey uint64
+		if e.expr.Grouping.Without {
+			groupingKey, buf = metric.HashWithoutLabels(buf, e.expr.Grouping.Groups...)
+		} else {
+			groupingKey, buf = metric.HashForLabels(buf, e.expr.Grouping.Groups...)
+		}
+
+		// TODO(karsten): support floats.
+		topkAggregation.ObserveForGroupingKey(s.Metric.String(), strconv.FormatUint(groupingKey, 10), uint32(s.F))
```

Comment:
So we should either use a number for the grouping key or the labels. I think the later is easier to debug.

Reply:
When I was running the tests I just saw the hash value as a string. It was not easy to identify each series. With the label string it's straight forward.

Reply:
For this PR we need to:

- [ ] Decide what the grouping key type should be?
- [ ] Add support for floats.

Reply:
I think we'd basically need to reimplement the `HashWithoutLabels` minus the last line: https://github.com/prometheus/prometheus/blob/main/model/labels/labels.go#L189-L208

As long as it's not a big perf. hit somehow I don't have a preference either way.